spring-cloud-gcp-vision/src/test/java/org/springframework/cloud/gcp/vision/it/DocumentOcrTemplateIntegrationTests.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.vision.it;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import com.google.cloud.vision.v1.TextAnnotation;
+import com.google.protobuf.InvalidProtocolBufferException;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.gcp.storage.GoogleStorageLocation;
+import org.springframework.cloud.gcp.vision.DocumentOcrResultSet;
+import org.springframework.cloud.gcp.vision.DocumentOcrTemplate;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.util.concurrent.ListenableFuture;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assumptions.assumeThat;
+
+@RunWith(SpringRunner.class)
+@ContextConfiguration(classes = { VisionTestConfiguration.class })
```

Comment:
Is it possible to rely on real configuration? It's nice when integration tests verify the way starter will be used in real life.

Reply:
Agreed, unfortunately I couldn't quite get this to work. Currently, the autoconfiguration project depends on the Vision project ([to provide beans of Vision templates](https://github.com/spring-cloud/spring-cloud-gcp/blob/master/spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/vision/CloudVisionAutoConfiguration.java#L85)). If I reference the real configuration here, I would have to make the vision project depend on autoconfiguration project which produces a circular dependency.

Do you have any suggestions? Maybe I am missing something with how Spring dependency injection works... In the Guice world I would only have the vision autoconfiguration provide `ImageAnnotatorClient` and be able to delete the autoconfiguration's dependency on the vision module. However, it seems like in Spring, I must declare the beans for the templates explicitly which forces the dependency between autoconfiguration and vision.

Reply:
Makes sense, thanks! This way we exercise both, the pure template usage, and the template+autoconfig usage in sample test.